feat(archive/100-theorems-list): Stirling

mo271 opened the pull request:
Co-authored-by: fabian-kruse <fabian_kruse@gmx.de>
Co-authored-by: nick-kuhn <46423253+nick-kuhn@users.noreply.github.com>



---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

mo271 committed:
feat(archive/100-theorems-list): Stirling

Co-authored-by: fabian-kruse <fabian_kruse@gmx.de>
Co-authored-by: nick-kuhn <46423253+nick-kuhn@users.noreply.github.com>

mo271 committed:
no dots in the authors lines

mo271 committed:
replace inv_pow₀ and ᾰ, don't import tactic

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+
+open_locale big_operators -- notation ∑ for finite sums
+open_locale classical real topological_space nnreal ennreal filter big_operators
+open real
```
```suggestion
open real finset nat filter
```

mo271 replied:
done

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+/-- **Sum of the Reciprocals of the Triangular Numbers**
+ (copied from archive)
+ TODO: include in some form mathlib or figure out how to import it from archive/ --/
+lemma inverse_triangle_sum :
```
This is something I usually say with lemmas like this, but it would be nice to instead have it as a lemma with the hypothesis `n ≠ 0`.

mo271 replied:
I agree. Here I just copied it verbatim from [archive/100-theorems-list/42_inverse_triangle_sum.lean](https://github.com/leanprover-community/mathlib/blob/master/archive/100-theorems-list/42_inverse_triangle_sum.lean). In order to use it I'm looking for a suitable place in `src`.

mo271 replied:
By now, this is not relevant anymore, because as @hrmacbeth suggested, we can use a more crude approximation in the proof...

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+Define `an n` as $\frac{n!}{\sqrt{2n}/(\frac{n}{e})^n$.
+Stirling's formula states that this sequence has limit $\sqrt(π)$.
+-/
+noncomputable def an (n : ℕ) : ℝ := (n.factorial : ℝ) / ((sqrt(2 * n) * ((n / (exp 1))) ^ n))
```
It's perhaps a bit too generic, but why not `a n` instead of `an n`?

vihdzp replied:
It would be a much longer name, but maybe you could call this `stirling_seq` and rename `bn` to `log_stirling_seq`?

mo271 replied:
I like the new names!

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+For any natural number `n`, the expression `log((n + 1) / n)` has the series expansion
+$\sum_{k=0}^{\infty}\frac{2}{2k+1}(\frac{1}{2n+1})^{2k+1}$.
+-/
+lemma power_series_ln (n : ℕ) (hn : n ≠ 0) : has_sum (λ (k : ℕ),
```
The lemma name probably shouldn't use `ln` since that's not how we call it in mathlib. Maybe `power_series_log_succ_div`?

mo271 replied:
done

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+For any natural number `n ≠ 0`, we have the identity
+`log((n + 1) / n) = log(1 + 1 / (2*n + 1)) - log(1 - 1/(2*n + 1))`.
+-/
+lemma aux_log (n : ℕ) (hn : n ≠ 0) :
```
Maybe `log_succ_div_eq_log_sub`?

mo271 replied:
done

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+/--
+`bn n` is log of `an n`.
+-/
+noncomputable def bn (n : ℕ) : ℝ := log (an n)
```
Same here, why not `b n` instead of `bn n`?

mo271 replied:
using `log_stirling_seq` instead of `bn` as you suggested above.

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+For each natural number `n ≠ 0`, we have $0<\sqrt{2n}$.
+-/
+lemma zero_lt_sqrt_two_n (n : ℕ) (hn : n ≠ 0) : 0 < sqrt (2 * (n : ℝ)) :=
+   real.sqrt_pos.mpr (mul_pos two_pos (cast_pos.mpr (zero_lt_iff.mpr hn)))
```
This shouldn't be indented. Also, is it not more convenient to just have a lemma `0 < sqrt (n : ℝ)`?

mo271 replied:
indentation done.
Still need to change the lemma

nick-kuhn replied:
Done. Deleted the Lemma and replaced occurences with a direct proof

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+end
+
+/-- The sequence `bn` is monotone decreasing -/
+lemma bn_antitone : ∀ (n m : ℕ), n ≤ m → bn m.succ ≤ bn n.succ :=
```
Why not prove `antitone bn` directly? There should be some sort of lemma somewhere in mathlib stating that it suffices that `bn n.succ ≤ bn n`.

mo271 replied:
The reason is that I use that form of antitone-ness (for an, now named stirling_seq) in `stirling_seq_has_limit_a` what `monotone_convergence` expects is exactly that.

nick-kuhn replied:
I changed the statement to `antitone (log_stirling_seq ∘ succ)`, which is definitionally equal. I believe the other point is already addressed. Also changed the name and documentation.

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+end
+
+/-- The sequence `bn` is bounded below. -/
+lemma bn_has_lower_bound : (lower_bounds (set.range (λ (k : ℕ), bn k.succ))).nonempty :=
```
You can use `bdd_below`.

mo271 replied:
When renaming and using `bdd_below`, I realized that this lemma is not even used, so I removed it.

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+Any sequence `bn` of real numbers that is monotone decreasing and bounded below has
+a limit in the real numbers.
+-/
+lemma monotone_convergence (bn : ℕ → ℝ) (h_sd : ∀ (n m : ℕ), n ≤ m → bn m ≤ bn n)
```
You'll probably find it easier to just use `infi bn` together with the API on it instead of using this theorem.

mo271 replied:
will look into that.

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+end
+
+/-- The sequence `an` is positive for `n > 0`  -/
+lemma an'_pos : ∀ (n : ℕ), 0 < an n.succ :=
```
You can use `$` to get rid of some parentheses. Also, why not put `(n : ℕ)` before the semicolon?

mo271 replied:
I didn't know how to make this a term proof with the `(n : ℕ)` before the semicolon.
Will check out how to use `$` here.

nick-kuhn replied:
Put the `(n : ℕ)` before the semicolon. 
Not familiar with using `$`.

alreadydone replied:
This is a maximally `$`-ified proof:
```
lemma stirling_seq'_pos (n : ℕ) : 0 < stirling_seq n.succ :=
div_pos (cast_pos.mpr $ factorial_pos n.succ : (0:ℝ) < _) $
  mul_pos (real.sqrt_pos.mpr $ mul_pos two_pos $ cast_pos.mpr $ succ_pos n) $
    pow_pos (div_pos (cast_pos.mpr $ succ_pos n) $ exp_pos 1) n.succ
```

The rule is to put $ before the last argument of the function and remove the parentheses enclosing the last argument.

alreadydone replied:
By the way, currently there are some indentations in `log_stirling_seq_bounded_aux`, `log_stirling_seq_bounded_by_constant`, `monotone_convergence` and `stirling_seq'_bdd_below` using an odd number of spaces, which should be fixed (always add two spaces from one indentation level to the next).

nick-kuhn replied:
Thanks, that helped! I got rid of some triple parentheses and the spacing should be fixed now.

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+end
+
+/-- The sequence `an` is monotone decreasing -/
+lemma an'_antitone : ∀ (n m : ℕ), n ≤ m → an m.succ ≤ an n.succ :=
```
You should directly prove `antitone an` instead.

mo271 replied:
see comment above, I don't really need `antitone an`, but a weaker statement. Perhaps I could change the name instead?

nick-kuhn replied:
I changed the statement to `antitone (stirling_seq ∘ succ)`, which is definitionally equal. Also changed the documentation. The name was already changed.

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+  (λ n, λ m, λ h, (log_le_log (an'_pos m) (an'_pos n)).mp (bn_antitone n m h))
+
+/-- The sequence `an` has a lower bound -/
+lemma an'_has_lower_bound : (lower_bounds (set.range (λ (k : ℕ), an k.succ))).nonempty :=
```
You can use `bdd_below` here too.

mo271 replied:
Done. Changed name and content

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+/--
+We can shift the indexing in a product.
+ -/
+lemma aux1 (k : ℕ) : ∏ i in range k, (wallis_inside_prod (1 + i)) =
```
Is this not a much more general theorem?

mo271 replied:
Yes, sure, it doesn't depend at all on the specific function. Perhaps this is something to include somewhere in the big_operators part of mathlib. (I hope I didn't miss it there...)

nick-kuhn replied:
It's basically `prod_Ico_add` and `range_eq_Ico` put together (which is how it was proven). Here, one has to fiddle around a bit to get the `n+1` changed to `succ n` inside the product. Since it was used exactly once, I deleted the lemma and inserted the proof at the relevant location.

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+end
+
+/-- For `n : ℕ`, define `wn n` as `2^(4*n) * n!^4 / ((2*n)!^2 * (2*n + 1))` -/
+noncomputable def wn (n : ℕ) : ℝ :=
```
Once again, why not `w n` instead of `wn n`?

mo271 replied:
changing to `w`

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+noncomputable def wn (n : ℕ) : ℝ :=
+  ((2 : ℝ) ^ (4 * n) * (n.factorial) ^ 4) / ((((2 * n).factorial) ^ 2) * (2 * (n : ℝ) + 1))
+
+/-- The sequence `wn n`converges to `π/2`-/
```
There's some weird spacing here and in subsequent comments.

mo271 replied:
improved spacing and math in comments

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+h.comp (tendsto_id.const_mul_at_top' two_pos)
+
+/-- For `n : ℕ`, define `cn n` as √2n * (n/e)^(4n) *2^(4n) / ((√4n (2n/e)^(2n))^2 * (2n+1)) -/
+noncomputable def cn (n : ℕ) : ℝ := ((sqrt (2 * n) * ((n / (exp 1)) ^ n)) ^ 4) * 2 ^ (4 * n) /
```
That same remark about `c n` here too.

mo271 replied:
done

vihdzp commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,874 @@
+  rw [one_div, div_eq_mul_inv],
+end
+
+/-- Stirling's Formula -/
```
```suggestion
/-- **Stirling's Formula** -/
```

mo271 replied:
done

mo271 committed:
review comments

mostly changes in naming,
one lemma eliminated

mo271 committed:
lint

nick-kuhn committed:
removed lemma zero_lt_sqrt_two_n

nick-kuhn committed:
Addressed some comments

nick-kuhn committed:
Removed aux1

nick-kuhn committed:
inserted $ to get rid of parentheses;
fixed some spacings

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,872 @@
+lemma stirling_seq'_pos (n : ℕ): 0 < stirling_seq n.succ :=
+  div_pos ((cast_pos.mpr (factorial_pos n.succ)): (0:ℝ) < _) $
+    mul_pos (real.sqrt_pos.mpr $ mul_pos two_pos $ cast_pos.mpr (succ_pos n)) $
+      pow_pos (div_pos (cast_pos.mpr (succ_pos n)) (exp_pos 1)) n.succ
```
```suggestion
begin
  apply_rules [cast_pos.mpr, factorial_pos, exp_pos, pow_pos, div_pos, mul_pos, real.sqrt_pos.mpr,
    two_pos, succ_pos];
  apply_instance
end
```
Perhaps this is more suggestive?  I've been toying with the idea of writing a `positivity` tactic for this kind of thing, cc @robertylewis

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,872 @@
+  tendsto (λ n : ℕ, (an n.succ)) at_top (𝓝 a) :=
+begin
+  nth_rewrite_rhs 0 ← tendsto_map'_iff,
+  have h : map succ at_top = at_top :=
```
This seems worthy of a separate lemma, maybe a simp-lemma
```lean
@[simp] lemma nat.map_succ_at_top : map (succ : ℕ → ℕ) at_top = at_top :=
```
It might even already exist, you could ask on "Is there code for X"?

mo271 replied:
https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/map.20.28succ.20.3A.20.E2.84.95.20.E2.86.92.20.E2.84.95.29.20at_top.20.3D.20at_top

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,872 @@
+sequence given by `an n.succ` has the limit `a`.
+-/
+lemma tendsto_succ (an : ℕ → ℝ) (a : ℝ) : tendsto an at_top (𝓝 a) ↔
+  tendsto (λ n : ℕ, (an n.succ)) at_top (𝓝 a) :=
```
Relocate and generalize from reals to
```lean
{α : Type*} (an : ℕ → α) (l : filter α)
```

mo271 replied:
Relocate to where? Somewhere in `analysis.asymptotics`?
Btw, we could think about reformulating Stirlings lemma using the relation `is_equivalent`.

mo271 replied:
Using `filter.tendsto_add_at_top_iff_nat` now...
Removing `tendsto_succ`

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,872 @@
+    exact gt_of_ge_of_gt ((le_is_glb_iff (is_glb_of_tendsto_at_top stirling_seq'_antitone ha)).mpr
+      e_lower_bound) (3 / 4 - 1 / 2 * log 2).exp_pos },
+  { exact ha },
+end
```
```suggestion
/--
The sequence `stirling_seq` has a positive lower bound (in fact, `exp (3/4 - 1/2 * log 2)`)
-/
lemma stirling_seq'_bounded_by_pos_constant :
  ∃ a, 0 < a ∧ ∀ n : ℕ, a ≤ stirling_seq n.succ :=
begin
  refine ⟨exp (3 / (4 : ℝ) - 1 / 2 * log 2), exp_pos _, λ n, _⟩,
  rw ← le_log_iff_exp_le (stirling_seq'_pos n),
  exact log_stirling_seq_bounded_by_constant n,
end

/-- The sequence `stirling_seq ∘ succ` is monotone decreasing -/
lemma stirling_seq'_antitone : antitone (stirling_seq ∘ succ) :=
λ n m h, (log_le_log (stirling_seq'_pos m) (stirling_seq'_pos n)).mp (log_stirling_seq'_antitone h)

/-- The limit `a` of the sequence `stirling_seq` satisfies `0 < a` -/
lemma stirling_seq_has_pos_limit_a :
  ∃ (a : ℝ), 0 < a ∧ tendsto (λ (n : ℕ), stirling_seq n) at_top (𝓝 a) :=
begin
  obtain ⟨x, x_pos, hx⟩ := stirling_seq'_bounded_by_pos_constant,
  have hx' : x ∈ lower_bounds (set.range (stirling_seq ∘ succ)) := by simpa [lower_bounds] using hx,
  refine ⟨_, lt_of_lt_of_le x_pos (le_cInf (set.range_nonempty _) hx'), _⟩,
  rw tendsto_succ,
  exact tendsto_at_top_cinfi stirling_seq'_antitone ⟨x, hx'⟩,
end
```
A shorter version (I hope it's instructive).  You can also get rid of your `monotone_convergence` lemma, which after all is basically just `tendsto_at_top_cinfi`.

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,872 @@
+    ... = -1 / 2 * log 2 + 1 - 1 / 4  : by simp only [one_div, log_inv, log_exp, sub_neg_eq_add]
+    ... = -1 / 2 * log 2 + 3 / 4      : by ring
+    ... = 3 / (4 : ℝ) - 1 / 2 * log 2 : by ring,
+end
```
In fact, you can go nonconstructive earlier.  It will suffice to make this
```lean
/-- The sequence `log_stirling_seq` is bounded below for `n ≥ 1`. -/
lemma log_stirling_seq_bounded_by_constant :
  ∃ c, ∀ n : ℕ, c ≤ log_stirling_seq n.succ :=
⟨_, λ n, sub_le.mp (log_stirling_seq_bounded_aux n)⟩
```

mo271 replied:
done

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,872 @@
+  norm_cast,
+  rw sq,
+  linarith,
+end
```
```suggestion
lemma log_stirling_seq_sub_log_stirling_seq_succ (n : ℕ) :
  log_stirling_seq n.succ - log_stirling_seq n.succ.succ ≤ 1 / (4 * n.succ * n.succ.succ) :=
begin
  have h₁ : 0 < 4 * ((n:ℝ) + 1) * (((n:ℝ) + 1) + 1) := by nlinarith [@cast_nonneg ℝ _ n],
  have h₃ : 0 < (2 * ((n:ℝ) + 1) + 1) ^ 2 := by linarith,
  have h₂ : 0 < 1 - (1 / (2 * ((n:ℝ) + 1) + 1)) ^ 2,
  { rw ← mul_lt_mul_right h₃,
    have H : 0 < (2 * ((n:ℝ) + 1) + 1) ^ 2 - 1 := by linarith,
    convert H using 1; field_simp [h₃.ne'] },
  refine le_trans (log_stirling_seq_diff_le_geo_sum n) _,
  push_cast at *,
  rw div_le_div_iff h₂ h₁,
  rw ← mul_le_mul_right h₃,
  have H : 4 * ((n:ℝ) + 1) * (n + 1 + 1) ≤ (2 * (↑n + 1) + 1) ^ 2 - 1 := by linarith,
  convert H using 1; field_simp [h₃.ne'],
end
```
Another golf.  The lesson here is that mathlib has very good tools for inequalities (`(n)linarith`) and for denominator-cancelling (`field_simp`), but can't really mix them.  It's fastest to reorganize so that the steps of the two types are kept separate.

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,872 @@
+      { refine sum_congr rfl _, intros k hk, exact h_pow_succ k },
+    end,
+    norm_num,
+    exact h_geom'',
```
I think you can just do
```suggestion
    exact (has_sum_geometric_of_lt_1 h_nonneg hlt).mul_left ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2)
```

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,872 @@
+  begin
+    intro k,
+    have h_zero_le : 0 ≤ ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) ^ k.succ, by
+    { simp [cast_succ, one_div, inv_pow, inv_nonneg], norm_cast, exact zero_le', },
```
```suggestion
    have h_zero_le : 0 ≤ ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) ^ k.succ := pow_nonneg h_nonneg _,
```
if you pull `h_nonneg` out from the proof of `g`.

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,872 @@
+  rw ← hx at h_sum,
+  exact (summable.has_sum_iff_tendsto_nat $
+    (summable_nat_add_iff 1).mpr $ has_sum.summable h_sum₁).mpr h_sum,
+end
```
```suggestion
begin
  change
    has_sum ((λ (b : ℕ), 1 / (2 * (b : ℝ) + 1) * ((1 / (2 * (m.succ : ℝ) + 1)) ^ 2) ^ b) ∘ succ) _,
  rw has_sum_nat_add_iff 1,
  convert (power_series_log_succ_div m.succ (succ_ne_zero m)).mul_left ((m.succ : ℝ) + 1 / (2 : ℝ)),
  { ext k,
    rw [← pow_mul, pow_add],
    have : 2 * (k : ℝ) + 1     ≠ 0, by {norm_cast, exact succ_ne_zero (2*k)},
    have : 2 * (m.succ :ℝ) + 1 ≠ 0, by {norm_cast, exact succ_ne_zero (2*m.succ)},
    field_simp,
    ring },
  { have h_reorder : ∀ {a b c d e f : ℝ}, a - 1 / (2 : ℝ) * b - c - (d - 1 / (2 : ℝ) * e - f) =
      (a - d) - 1 / (2 : ℝ) * (b - e) - (c - f),
    by {intros, ring_nf},
    rw [log_stirling_seq_formula, log_stirling_seq_formula, h_reorder],
    repeat {rw [log_div, factorial_succ]},
    push_cast,
    repeat {rw log_mul},
    rw log_exp,
    ring_nf,
    all_goals {norm_cast},
    all_goals {try {refine mul_ne_zero _ _}, try {exact succ_ne_zero _}},
    any_goals {exact factorial_ne_zero m},
    any_goals {exact exp_ne_zero 1},
    simp },
  { apply_instance }
end
```
Two tricks to this golf:
- finding the right lemma (`has_sum_nat_add_iff`)
- reorganizing to avoid the long sequence of transformations `h_sum₀`, `h_sum₁`, `h_sum` -- instead match to something broadly correct using `convert` and then you just have equality goals, which are easier.

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,872 @@
+
+/-- **Sum of products of consecutive reciprocals** -/
+lemma partial_sum_consecutive_reciprocals :
+ ∀ n, ∑ k in range n, (1 : ℚ) / (k.succ * (k.succ.succ)) ≤ 1 :=
```
These lemmas about triangular numbers are nice, but I think you could defer them to a separate project.  All you need for this project is the summability of say `1 / (k.succ)^2`, right?  (without a precise knowledge of what the sum is).  And this is already in the library as [real.summable_one_div_nat_rpow](https://leanprover-community.github.io/mathlib_docs/analysis/p_series.html#real.summable_one_div_nat_rpow).

mo271 replied:
That makes a lot of sense! I removed the lemma about the triangular numbers and instead base it on the fact that all partial sums of zeta(2) bounded. (Left this as a sorry for now..)

mo271 replied:
sorry closed, but it is still a bit clumsy...

hrmacbeth replied:
Thanks for the change!  I streamlined a little in my golf below.

mo271 committed:
Apply suggestions from code review

Co-authored-by: Heather Macbeth <25316162+hrmacbeth@users.noreply.github.com>

mo271 committed:
h_nonneg out of scope

mo271 committed:
use c instead of concrete constant

mo271 committed:
h_nonneg out of scope

mo271 committed:
using filter.tendsto_add_at_top_iff_nat

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,729 @@
+/-- The sequence `log_stirling_seq` is bounded below for `n ≥ 1`. -/
+lemma log_stirling_seq_bounded_by_constant : ∃ c, ∀ (n : ℕ), c ≤ log_stirling_seq n.succ :=
+begin
+  use 3 / (4 : ℝ) - 1 / 2 * log 2,
```
My point here is that in fact you can 
```suggestion
  use log_stirling_seq 1 - 1 / 4,
```
which shortens the proof a lot!

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,729 @@
+begin
+  use Inf (set.range b),
+  exact tendsto_at_top_is_glb h_sd (real.is_glb_Inf (set.range b) (set.range_nonempty b) h_bounded),
+end
```
Don't need this lemma any more.

mo271 replied:
done

mo271 committed:
using less concrete bounds

mo271 committed:
remove monotone_convergence

mo271 committed:
towards not using \le 1

mo271 committed:
remove triangular numbers lemmas

mo271 committed:
rename lemma

mo271 committed:
closing last sorry

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,718 @@
+lemma log_stirling_seq_bounded_by_constant : ∃ c, ∀ (n : ℕ), c ≤ log_stirling_seq n.succ :=
+begin
+  have h := log_stirling_seq_bounded_aux,
+  cases h with d h,
```
This pattern (which appears several times in your code, please change everywhere) can be shortened to
```suggestion
  cases log_stirling_seq_bounded_aux with d h,
```
or
```suggestion
  obtain ⟨d, h⟩ := log_stirling_seq_bounded_aux,
```

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,718 @@
+We have the bound  `log_stirling_seq n - log_stirling_seq (n+1)` ≤ 1/(4n(n+1))
+-/
+lemma log_stirling_seq_sub_log_stirling_seq_succ (n : ℕ) :
+  log_stirling_seq n.succ - log_stirling_seq n.succ.succ ≤ 1 / (4 * n.succ * n.succ.succ) :=
```
A general principle is to claim only what you need, as early as possible.  This saves effort later.  So here I'd switch the RHS of the lemma statement to ```1 / (4 * n.succ ^ 2)```.

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,718 @@
+     refine (mul_le_mul_left _).mpr _, { exact one_div_pos.mpr four_pos, },
+     exact h n,
+    end
+end
```
With the change I propose to the previous lemma, and some golfing (following the principle "reorganize to do exactly one thing at a time") this becomes
```suggestion
begin
  let d := ∑' k : ℕ, (1 : ℝ) / (k.succ)^2,
  use (1/4 * d : ℝ),
  let log_stirling_seq' : (ℕ → ℝ) := λ (k : ℕ), log_stirling_seq k.succ,
  intro n,
  calc
  log_stirling_seq 1 - log_stirling_seq n.succ = log_stirling_seq' 0 - log_stirling_seq' n : rfl
    ... = ∑ k in range n, (log_stirling_seq' k - log_stirling_seq' (k + 1)) :
    by rw ← (sum_range_sub' log_stirling_seq' n)
    ... ≤ ∑ k in range n, (1/4) * (1 / (k.succ)^2) : 
    begin
      apply sum_le_sum,
      intros k hk,
      convert log_stirling_seq_sub_log_stirling_seq_succ k using 1,
      field_simp,
    end
    ... = 1 / 4 * ∑ k in range n, 1 / (k.succ)^2 : by rw mul_sum
    ... ≤ 1 / 4 * d :
    begin
      refine (mul_le_mul_left _).mpr _, { exact one_div_pos.mpr four_pos, },
      refine sum_le_tsum (range n) (λ k _, _) summable_inverse_squares,
      apply le_of_lt,
      rw one_div_pos,
      rw sq_pos_iff,
      exact nonzero_of_invertible ↑(succ k)
    end
end
```

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,718 @@
+
+
+lemma partial_sum_le_tsum {a: ℕ → ℝ} (h : summable a) (g: ∀ (n:ℕ), 0 ≤ a n):
+∃ (c:ℝ), ∀ (s : finset ℕ), s.sum a ≤ c :=
```
This lemma is a bit too specialized.  I eliminated the need for it in my golf below.

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,718 @@
+
+/-- **Partial sum inverses of squares** -/
+lemma partial_sum_inverse_squares :
+ ∃ c, ∀ n, ∑ k in range n, (1 : ℝ) / (k.succ)^2 ≤ c :=
```
This lemma is a bit too specialized.  I eliminated the need for it in my golf below.

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,718 @@
+-/
+lemma power_series_log_succ_div (n : ℕ) (hn : n ≠ 0) : has_sum (λ (k : ℕ),
+  (2 : ℝ) * (1 / (2 * (k : ℝ) + 1)) * ((1 / (2 * (n : ℝ) + 1)) ^ (2 * k + 1)))
+  (log ((n.succ : ℝ) / (n : ℝ))) :=
```
This lemma (and the two previous) hold for positive real `n`, not just natural numbers `n`.  Stating it in this way will postpone the casting difficulties until they are really relevant, and it will also make this lemma a better candidate for inclusion in the main library, near `has_sum_pow_div_log_of_abs_lt_1`.

(I'd probably rename from `n` to something more suggestive of a real number.)

mo271 replied:
Done (it only touched this lemma and the previous one, in the one before that we already had everything real)

I you think it is a good idea, I can factor out `power_series_log_succ_div` and it's two prerequisites into `analysis.special_functions.log.deriv`.

hrmacbeth commented on archive/100-theorems-list/90_stirlings_formula.lean:
```diff
@@ -0,0 +1,718 @@
+For each natural number `n ≠ 0`, we have $0<\sqrt{2n}$.
+-/
+lemma zero_lt_sqrt_two_n (n : ℕ) (hn : n ≠ 0) : 0 < sqrt (2 * (n : ℝ)) :=
+real.sqrt_pos.mpr $ mul_pos two_pos $ cast_pos.mpr (zero_lt_iff.mpr hn)
```
Doesn't seem to be used?

nick-kuhn committed:
Following comments:

Fixed wasteful have-cases pattern;
Improved log_stirling_seq_sub_log_stirling_seq_succ  and
log_stirling_seq_bounded_aux

Co-authored-by: Heather Macbeth <25316162+hrmacbeth@users.noreply.github.com>

nick-kuhn committed:
Addressed coments by hrmacbeth

mo271 committed:
more general log lemmas

mo271 committed:
lint
